sympy/sets/tests/test_fancysets.py

```diff
@@ -236,24 +237,24 @@ def test_Range_set():
     assert Range(2, -oo, -2)[:2:2] == Range(2, -2, -4)
     raises(ValueError, lambda: Range(-oo, 4, 2)[:2:2])
     assert Range(-oo, 4, 2)[::-2] == Range(2, -oo, -4)
-    raises(ValueError, lambda: Range(-oo, 4, 2)[::2])
```

Comment:
Why was there previously a ValueError?

Reply:
I think I see...as long as a slice doesn't involve counting from the infinite end then the slice can be computed since we know where it ends (e.g. `Range(-oo, 3)[-1] == 2`).

Reply:
I think this should raise. In fact it should probably raise without being sliced.

Reply:
I agree with @smichr , because, by doing something like, `[::2]` we are just making changing the step so there is no harm in doing it. 
In fact, @smichr is right IMO because, until we are not accessing the end points, we should not worry about whether they are `oo` or `-oo`.
@Upabjojr @sidhantnagpal Any comments on this from your side.

Reply:
And we have a convention for the non-infinite end: when at the start, that is where we start. When at the end, the last emitted value is `end - step`.

Reply:
This Range should just raise but we don't know what any of its elements are. There's no way to sensibly define a range with an infinite start unless the step is +-1. The oo start/end of the Range needs to interpreted as some sort of limit but but there isn't an unambiguous limit in these cases:
```julia
In [8]: n = Symbol('n', integer=True, positive=True)                                                     

In [9]: r1 = Range(-2*n, 1, 2)                                                                           

In [10]: r2 = Range(-2*n-1, 1, 2)                                                                        

In [11]: 0 in r1                                                                                         
Out[11]: True

In [12]: 0 in r2                                                                                         
Out[12]: False

In [13]: r1.subs(n, oo)                                                                                  
Out[13]: {…, -3, -1}

In [14]: r2.subs(n, oo)                                                                                  
Out[14]: {…, -3, -1}

In [15]: r1.subs(n, oo) == r2.subs(n, oo)                                                                
Out[15]: True
```

Reply:
So, as far as I can figure out, you meant to say that `Range(-oo, 4, 2)` shouldn't have been created in the first place, because there is an ambiguity in it's elements. Well, the experiment you did is quite valid. It is making me think that here slicing is not issue instead, it is about the `__new__` method probably some changes are required to be made and this thing has to be taken into account. `Range` shouldn't start with `-oo` or `oo` if the step is not +/-1. @smichr Any comments?

Reply:
> There's no way to sensibly define a range with an infinite start unless the step is +-1

It is ok to start with infinite values. This is already addressed in the docstring of master:

> Infinite ranges are allowed. ``oo`` and ``-oo`` are never included in the
 set (``Range`` is always a subset of ``Integers``). If the starting point
 is infinite, then the final value is ``stop - step``. To iterate such a
 range, it needs to be reversed:

Reply:
So, no problems with the new results of slicing. May be, after applying(locally on my system I have applied) https://github.com/sympy/sympy/pull/17146#issuecomment-515649370, slicing can be done with `oo` or `-oo` at an end because, accessing any element with `oo` or `-oo` as the starting point will return `oo` or `-oo`. Is it fine?

Reply:
Please show an example.

Regarding the printed value for sup. Maybe the following would be useful:
```python
def abcis(*vi):
    # return canonical a, b, c, inf, sup or (nan,)*5
    # if Piecewise are returned, if `a` becomes nan after
    # substitution, all values can be taken to be nan
    x = a,b,c = symbols('a,b,c')
    c1 = Or((b - a)*c <= 0, Eq(1/c, 0), Eq(1/a, 1/b))
    c2 = Or(Eq(a, oo), Eq(b, oo))
    c3 = Or(Eq(a, -oo), Eq(b, -oo))
    c4 =  And(Eq(frac(a), 0), Eq(frac(b), 0), Eq(frac(c), 0))
    d = a + c*(ceiling((b - a)/c) - 1)
    e = Piecewise((b - c, Eq(1/a, 0)), (a, True))
    A = Piecewise(
        (nan, c1),
        (e, c2),
        (-oo, c3), (Piecewise((Piecewise(
        (a, b > a),
        (d, True)), c4),
        (nan, True)), True))
    B = Piecewise(
        (nan, c1),
        (oo, c2),
        (e, c3), (Piecewise((Piecewise(
        (d, b > a),
        (a, True)), c4),
        (nan, True)), True))
    r = dict(zip(x,vi))
    A = A.subs(r)
    if A is nan:
        return (nan,)*5
    B = B.subs(r)
    C = vi[-1]
    return Piecewise(
        ((A, B + C, C, A, B), C > 0),
        ((B, A + C, C, A, B),True))
```
Tested and demo with
```python
v=[(1, 3, 2), (1, 4, 2), (1, oo, 2), (-oo, 1, 2), (1, 1, 2), (1, 1, 0), (-oo, oo, 2)]
v+=[(j,i,-k) for i,j,k in v]
x=a,b,c
for vi in v:
    ans = abcis(*vi)
    print(vi,ans)
    if ans[0] is not S.NaN:
        r = Range(*vi)
        assert r.args+(r.inf, r.sup) == ans

(1, 3, 2) (1, 3, 2, 1, 1)
(1, 4, 2) (1, 5, 2, 1, 3)
(1, oo, 2) (1, oo, 2, 1, oo)
(-oo, 1, 2) (-oo, 1, 2, -oo, -1)
(1, 1, 2) (nan, nan, nan, nan, nan)
(1, 1, 0) (nan, nan, nan, nan, nan)
(-oo, oo, 2) (nan, nan, nan, nan, nan)
(3, 1, -2) (3, 1, -2, 3, 3)
(4, 1, -2) (4, 0, -2, 2, 4)
(oo, 1, -2) (oo, 1, -2, 3, oo)
(1, -oo, -2) (1, -oo, -2, -oo, 1)
(1, 1, -2) (nan, nan, nan, nan, nan)
(1, 1, 0) (nan, nan, nan, nan, nan)
(oo, -oo, -2) (nan, nan, nan, nan, nan)
>>> abcis(1,b,3)
(Piecewise((nan, b <= 1), (1, Eq(b, oo)), (-oo, Eq(b, -oo)), (Piecewise((Piecewise((1, b > 1), (3*ceiling(b/3 - 1/3) - 2, True)), Eq(frac(b), 0)), (nan, True)), True)), Piecewise((nan, b <= 1), (oo, Eq(b, oo)), (1, Eq(b, -oo)),(Piecewise((Piecewise((3*ceiling(b/3 - 1/3) - 2, b > 1), (1, True)), Eq(frac(b), 0)), (nan, True)), True)) + 3, 3, Piecewise((nan, b <= 1), (1, Eq(b, oo)), (-oo, Eq(b, -oo)), (Piecewise((Piecewise((1, b > 1), (3*ceiling(b/3 - 1/3) - 2, True)), Eq(frac(b), 0)), (nan, True)), True)), Piecewise((nan, b <=1), (oo, Eq(b, oo)), (1, Eq(b, -oo)), (Piecewise((Piecewise((3*ceiling(b/3 - 1/3) - 2, b > 1), (1, True)), Eq(frac(b), 0)), (nan, True)), True)))

>>> abcis(1,var('b',integer=1),3)
(Piecewise((nan, b <= 1), (1, Eq(b, oo)), (-oo, Eq(b, -oo)), (Piecewise((1,b > 1), (3*ceiling(b/3 - 1/3) - 2, True)), True)), Piecewise((nan, b <= 1),(oo, Eq(b, oo)), (1, Eq(b, -oo)), (Piecewise((3*ceiling(b/3 - 1/3) - 2, b >1), (1, True)), True)) + 3, 3, Piecewise((nan, b <= 1), (1, Eq(b, oo)), (-oo, Eq(b, -oo)), (Piecewise((1, b > 1), (3*ceiling(b/3 - 1/3) - 2, True)), True)), Piecewise((nan, b <= 1), (oo, Eq(b, oo)), (1, Eq(b, -oo)), (Piecewise((3*ceiling(b/3 - 1/3) - 2, b > 1), (1, True)), True)))

>>> piecewise_fold(_[-1])  # the sup for Range(1, b, 3) with b and integer
Piecewise((nan, b <= 1), (oo, Eq(b, oo)), (1, Eq(b, -oo)), (3*ceiling(b/3 -1/3) - 2, True))
```

Reply:
I have also used a similar approach. AFAICT, you are using `S.NaN` for empty range. I will surely push the changes and will show an example.

Reply:
OK, so you are looking into why the inf value showing in the `{...}` is incorrect. It would be good to centralize the logic -- apparently it isn't thus now because this case has slipped through.

Regarding nan...I don't know how you can handle this with the dot-dot representation. Basically if the code I gave indicates nan for `a` that means you would have an empty set. So if we read `{1, 2, ... i}` as empty when `i < 1` then instead of nan you could use something to indicate such a number. For Range(1, b, 3) it would be {1, 4, ..., c} where `c = Piecewise((0, b <= 1), (oo, Eq(b, oo)), (3*ceiling(b/3 -1/3) - 2, True))`.